src/audio/codec_adapter/codec/generic.c

```diff
@@ -0,0 +1,366 @@
+		ret = -EIO;
+		goto out;
+	}
+	/* Assign interface */
+	codec->ops = interface;
+	/* Now we can proceed with codec specific initialization */
+	ret = codec->ops->init(dev);
+	if (ret) {
+		comp_err(dev, "codec_init() error %d: codec specific init failed, codec_id %x",
+			 ret, codec_id);
+		goto out;
+	}
+
+	comp_info(dev, "codec_init() done");
+	codec->state = CODEC_INITIALIZED;
+out:
+	return ret;
+}
+
+void *codec_allocate_memory(struct comp_dev *dev, uint32_t size,
+			    uint32_t alignment)
+{
+	struct codec_memory *container;
+	void *ptr;
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	if (!size) {
+		comp_err(dev, "codec_allocate_memory: requested allocation of 0 bytes.");
+		return NULL;
+	}
+
+	/* Allocate memory container */
+	container = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+			    sizeof(struct codec_memory));
+	if (!container) {
+		comp_err(dev, "codec_allocate_memory: failed to allocate memory container.");
+		return NULL;
+	}
+
+	/* Allocate memory for codec */
+	if (alignment)
+		ptr = rballoc_align(0, SOF_MEM_CAPS_RAM, size, alignment);
+	else
+		ptr = rballoc(0, SOF_MEM_CAPS_RAM, size);
+
+	if (!ptr) {
+		comp_err(dev, "codec_allocate_memory: failed to allocate memory for codec %x.",
+			 cd->ca_config.codec_id);
+		return NULL;
+	}
+	/* Store reference to allocated memory */
+	container->ptr = ptr;
+	list_item_prepend(&container->mem_list, &cd->codec.memory.mem_list);
+
+	return ptr;
+}
+
+int codec_free_memory(struct comp_dev *dev, void *ptr)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct codec_memory *mem;
+	struct list_item *mem_list;
+	struct list_item *_mem_list;
+
+	if (!ptr) {
+		comp_err(dev, "codec_free_memory: error: NULL pointer passed.");
+		return -EINVAL;
+	}
+	/* Find which container keeps this memory */
+	list_for_item_safe(mem_list, _mem_list, &cd->codec.memory.mem_list) {
+		mem = container_of(mem_list, struct codec_memory, mem_list);
+		if (mem->ptr == ptr) {
+			rfree(mem->ptr);
+			list_item_del(&mem->mem_list);
+			rfree(mem);
+			return 0;
+		}
+	}
+
+	comp_err(dev, "codec_free_memory: error: could not find memory pointed by %p",
+		 (uint32_t)ptr);
+
+	return -EINVAL;
+}
+
+static int validate_config(struct codec_config *cfg)
+{
+	//TODO: validation of codec specific setup config
+	return 0;
+}
+
+int codec_prepare(struct comp_dev *dev)
+{
+	int ret;
+	struct comp_data *cd = comp_get_drvdata(dev);
+	uint32_t codec_id = cd->ca_config.codec_id;
+	struct codec_data *codec = &cd->codec;
+
+	comp_dbg(dev, "codec_prepare() start");
+
+	if (cd->codec.state == CODEC_PREPARED)
+		return 0;
+	if (cd->codec.state < CODEC_INITIALIZED)
+		return -EPERM;
+
+	ret = codec->ops->prepare(dev);
+	if (ret) {
+		comp_err(dev, "codec_prepare() error %d: codec specific prepare failed, codec_id 0x%x",
+			 ret, codec_id);
+		goto end;
+	}
+
+	codec->s_cfg.avail = false;
+	codec->r_cfg.avail = false;
+	codec->r_cfg.data = NULL;
+	list_init(&codec->memory.mem_list);
+
+	if (codec->r_cfg.data)
```

Comment:
Why do we free cfg data here, can we have a comment.

Reply:
Yes, cfg has been applied during the prepare procedure so we no longer need it. Comment added as well.